### PR TITLE
warning when your line endings are changed

### LINKED ISSUE #1859: line endings changes reported by Git are not displayed in diff
Please summarise the issue in the title, and then use the template below to
fill out the details so we can reproduce the issue on our end.

### Description

Windows GitHub Desktop shows 282 file changes. 

### Version

**GitHub Desktop version:** 0.5.9

**OS version:** Windows 10

### Steps to Reproduce

I have a couple files that are different. It shows that hundreds are different. Further, each file listed as "different" shows NO DIFFERENCES.

Further, I downloaded the ZIP from git (into another folder), and extracted that in place... and it shows 292 file differences. All these files are the same because I just downloaded them from the repository that they belong to.(!)!)!???) So I really don't understand. How can GitHub Windows think that files I downloaded from the repository are different than what's in the repository?

This application is completely confusing my development process.

**Expected behaviour:** I'm shown the actual file differences. This worked yesterday on a different machine.

**Actual behaviour:** It shows that basically every file is different

**Reproduces how often:** Always

### Logs

There may be some relevant information in log files generated by GitHub
Desktop:

 - If you are on macOS, attach the most recent log file from:
   `~/Library/Application Support/GitHub Desktop/logs/*.desktop.production.log`
 - If you are on Windows, attach the most recent log file from:
   `%APPDATA%\\GitHub Desktop\\logs\\*.desktop.production.log`

The log files are organized by date, so see if anything was generated for
today's date.

#### Additional Information

Any additional information, configuration or data that might be necessary to
reproduce the issue.

### PR DESCRIPTION
**Pull request by shiftkey**
~~:cactus: will need to be refactored once #1893 lands :cactus:~~

Fixes #1859 in two ways:

 - display something when Git returns an empty diff for the file
 
screen shot 2017-06-06 at 12 42 44 pm

 - in the file header, display the `warning: (CRLF|CR|LF) will be replaced by (CRLF|CR|LF) in [path]` message to indicate there's a related change.

This is a partial solution - we don't actually help them get back to a happy place - and I'm open to design/UX feedback on making this experience nicer in general while we're in here.

 - [x] add tests to capture regressions
 - [x] review diff abstractions and ensure things all make sense

### CODE REVIEW

**Comment by shiftkey** [comment]
Opening this up to get some UX feels as well as a review 🍺

File: app/src/lib/git/spawn.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import { GitProcess } from 'dugite'
+
+type ProcessOutput = {
+  output: Buffer,
+  error: Buffer,
+}
+
+export function spawnAndComplete(args: string[], path: string, name: string, successExitCodes?: Set<number>): Promise<ProcessOutput> {
```
Comment (joshaber) [inline app/src/lib/git/spawn.ts]: Could you 📖 this?

File: app/src/models/diff.ts [hunk 2]
```diff
@@ -24,13 +24,35 @@ export enum DiffLineType {
...
 }
 
+type LineEnding = 'CR' | 'LF' | 'CRLF'
+
+export type LineEndingsChange = {
+  from: LineEnding,
+  to: LineEnding,
+}
+
+export function parseLineEndingText(text: string): LineEnding | null {
```
Comment (joshaber) [inline app/src/models/diff.ts]: 📖 here too

File: app/src/ui/changes/changed-file-details.tsx [hunk 3]
```diff
@@ -16,12 +18,25 @@ export class ChangedFileDetails extends React.Component<IChangedFileDetailsProps
     const status = this.props.status
     const fileStatus = mapStatus(status)
 
+    let metadataElement: JSX.Element | undefined
+    const diff = this.props.diff
+    if (diff.kind === DiffType.Text) {
+      if (diff.lineEndingsChange) {
+        const message = `Warning: line endings change from '${diff.lineEndingsChange.from}' to '${diff.lineEndingsChange.to}'.`
```
Comment (joshaber) [inline app/src/ui/changes/changed-file-details.tsx]: Should this be past tense? "change**d**"? I'm not sure 🤔

**Comment by joshaber** [comment]
>This is a partial solution - we don't actually help them get back to a happy place - and I'm open to design/UX feedback on making this experience nicer in general while we're in here.

What does getting them back to a happy place look like?

I think this PR is a strict improvement, but we could open another issue for whatever other actions we should do.

File: app/src/ui/changes/changed-file-details.tsx (hunk 3 above)
Comment (shiftkey) [inline app/src/ui/changes/changed-file-details.tsx]: I'll try both, but "Warning: line endings **have changed** from .." might be the best fit here.

**Comment by shiftkey** [comment]
> What does getting them back to a happy place look like?

Probably normalizing the line endings in the repository, like this well-worn resource for Windows repositories: https://help.github.com/articles/dealing-with-line-endings/#refreshing-a-repository-after-changing-line-endings

To make this more friendly, we could do a couple of things in a workflow-esque experience:

 - create a `.gitattributes` file in the repository with some sane defaults like `* text=auto` (this was the root cause of #1859)
 - normalize the line endings to clean things up, creating a new commit containing all the necessary changes

**Comment by shiftkey** [comment]
> I think this PR is a strict improvement, but we could open another issue for whatever other actions we should do.

Agreed. If those tasks :point_up: seem worth capturing and discussing further I'll spin off a new issue tomorrow.

**Comment by joshaber** [comment]
As usual, I'll defer to you and others who know the Windows world better 😬

**Comment by t7m** [comment]
The code is committed, but there's no new GitHub Desktop Download, is there? How does one get notified when that is updated?

**Comment by shiftkey** [comment]
> The code is committed, but there's no new GitHub Desktop Download, is there? 

We just did an update yesterday. We'll probably do an update early next week once we've got a few more PRs merged.

>  How does one get notified when that is updated?

The app will check for updates in the background and prompt you to restart when it has the update ready to install.

**Comment by stuartpb** [comment]
I'm confused: I've added a .gitattributes with `* text=auto` to my repository, and I'm still seeing a `Warning: line endings have changed from 'LF' to 'CRLF'.` warning triangle. Does that mean I'm about to commit the file, which has LF line endings in the repo, with CRLF line endings if I commit this change? Is it trying to tell me that `autocrlf` is not in effect for some reason? (And, if so, why not?)

The only time I want to be *warned* about CRLF translation is when it's *not happening*, ie. when I'm about to commit a file *with CRLF line endings* to the repo. From the sounds of this change as described in this thread, the triangle is trying to warn me that `autocrlf` is *going to apply*: why would I want to be *confronted* about that? It's *the behavior I explicitly requested*.

**Pull request merged**